controller/src/beerocks/master/db/db.cpp

```diff
@@ -2117,6 +2117,324 @@ bool db::get_hostap_is_acs_enabled(std::string mac)
+
+    if (dwell_time_msec <= 0) {
+        LOG(ERROR) << __FUNCTION__ << "Only >0 dwell time is supported!";
+        return false;
+    }
+
+    auto scan_config =
+        (single_scan) ? n->hostap->single_scan_config : n->hostap->continuous_scan_config;
+    scan_config->dwell_time_msec = dwell_time_msec;
+
+    return true;
+}
+
+int db::get_channel_scan_dwell_time_msec(const std::string &mac, const bool single_scan)
+{
+    auto n = get_node(mac);
+
+    if (!n) {
+        LOG(ERROR) << "node not found.... ";
+        return -1;
+    } else if (n->get_type() != beerocks::TYPE_SLAVE || n->hostap == nullptr) {
+        LOG(ERROR) << __FUNCTION__ << "node " << mac << " is not a valid hostap!";
+        return -1;
+    }
+
+    auto scan_config =
+        (single_scan) ? n->hostap->single_scan_config : n->hostap->continuous_scan_config;
+    return scan_config->dwell_time_msec;
+}
+
+bool db::set_channel_scan_pool(const std::string &mac, const std::set<uint8_t> &channel_pool,
+                               const bool single_scan)
+{
+    auto n = get_node(mac);
+    if (!n) {
+        LOG(ERROR) << "node not found.... ";
+        return false;
+    } else if (n->get_type() != beerocks::TYPE_SLAVE || n->hostap == nullptr) {
+        LOG(ERROR) << __FUNCTION__ << "node " << mac << " is not a valid hostap!";
+        return false;
+    }
+
+    //Validate new dcs channel pool with hostap_supported_channels list
+    //This is a workaround
+    static uint8_t supported_channels_array[] = {
+        1,  2,   3,   4,   5,   6,   7,   8,   9,   10,  11,  36,  40,  44,  48,  52,  56,  60,
+        64, 100, 104, 108, 112, 116, 120, 124, 128, 132, 136, 140, 144, 149, 153, 157, 161, 165};
+    static size_t supported_channels_array_size =
+        sizeof(supported_channels_array) / sizeof(uint8_t);
+    std::set<uint8_t> supported_channels(supported_channels_array,
+                                         supported_channels_array + supported_channels_array_size);
+    for (auto channel : channel_pool) {
+        if (supported_channels.find(channel) == supported_channels.end()) {
+            //even if one channel is not supported reject all pool
+            LOG(ERROR) << __FUNCTION__ << "channel =" << int(channel)
+                       << " is invalid, setting channel pool failed !";
+            return false;
+        }
+    }
+    auto scan_config =
+        (single_scan) ? n->hostap->single_scan_config : n->hostap->continuous_scan_config;
+    scan_config->channel_pool = std::set<uint8_t>(channel_pool); // creates a copy of the pool
+    LOG(DEBUG) << __FUNCTION__ << ", " << (single_scan ? "single" : "continuous")
+               << " scan, setting channel pool succeeded!";
+
+    return true;
+}
+
+const std::set<uint8_t> &db::get_channel_scan_pool(const std::string &mac, const bool single_scan)
+{
+    auto n = get_node(mac);
+    if (!n) {
+        LOG(ERROR) << "node not found.... ";
+        return *(std::make_shared<std::set<uint8_t>>());
```

Comment:
Does this work? AFAICS this has the same problem as `return std::set<uint8_t>();` only the compiler doesn't complain about it...

Reply:
OMG.

Reply:
This is basically creating a shared_ptr, then instead of returning the shared_ptr itself which would increment the reference count preventing the memory from being released, we return the contents and when we return they are invalid.

Reply:
Added a dummy_return, if we decide that using a dummy_return is incorrect we'll fix it in #665